source/common/http/conn_manager_utility.cc

```diff
@@ -131,6 +132,68 @@ void ConnectionManagerUtility::mutateRequestHeaders(Http::HeaderMap& request_hea
   if (config.tracingConfig()) {
     Tracing::HttpTracerUtility::mutateHeaders(request_headers, runtime);
   }
+  mutateXfccRequestHeader(request_headers, connection, config);
+}
+
+void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_headers,
+                                                       Network::Connection& connection,
+                                                       ConnectionManagerConfig& config) {
+  // When AlwaysForwardOnly is set, always forward the XFCC header without modification.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::AlwaysForwardOnly) {
+    return;
+  }
+  // When Sanitize is set, or the connection is not mutual TLS, remove the XFCC header.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::Sanitize ||
+      !(connection.ssl() && connection.ssl()->peerCertificatePresented())) {
+    request_headers.removeForwardedClientCert();
+    return;
+  }
+
+  // When ForwardOnly is set, always forward the XFCC header without modification.
+  if (config.forwardClientCert() == Http::ForwardClientCertType::ForwardOnly) {
+    return;
+  }
+
+  // TODO(myidpt): Handle the special characters in By and SAN fields.
+  std::vector<std::string> client_cert_details;
```

Comment:
Yeah, the join will be slower. My main aim in suggesting this was to avoid the repeated boilerplate pattern in the code. I think it could also have been made an inline function or something like that as well.